alchemist/alchemist-euclidean-geometry/src/main/kotlin/it/unibo/alchemist/model/implementations/geometry/euclidean2d/Segment2DImpl.kt

```diff
@@ -1,11 +1,139 @@
+/*
+ * Copyright (C) 2010-2020, Danilo Pianini and contributors
+ * listed in the main project's alchemist/build.gradle.kts file.
+ *
+ * This file is part of Alchemist, and is distributed under the terms of the
+ * GNU General Public License, with a linking exception,
+ * as described in the file LICENSE in the Alchemist distribution's top directory.
+ */
+
 package it.unibo.alchemist.model.implementations.geometry.euclidean2d
 
+import it.unibo.alchemist.fuzzyIn
 import it.unibo.alchemist.model.interfaces.geometry.Vector2D
+import it.unibo.alchemist.model.interfaces.geometry.euclidean2d.Intersection2D
+import it.unibo.alchemist.model.interfaces.geometry.euclidean2d.Line2D
 import it.unibo.alchemist.model.interfaces.geometry.euclidean2d.Segment2D
+import it.unibo.alchemist.rangeFromUnordered
+import org.danilopianini.lang.MathUtils.fuzzyEquals
+import java.lang.UnsupportedOperationException
 
 /**
- * Defines a segment from [first] to [second] in an euclidean bidimensional space.
+ * A [Segment2D] partly delegated to [Line2D]. Doubles are only compared with [fuzzyEquals]. Properties are computed
+ * only once upon creation.
  */
 data class Segment2DImpl<P : Vector2D<P>>(override val first: P, override val second: P) : Segment2D<P> {
-    override fun copyWith(first: P, second: P): Segment2D<P> = copy(first, second)
```

Comment:
Everything that may have different implementations, with different trade offs between readability and performance, I put in `Segment2DImpl`. What's the advantage of having most of the stuff default-implemented?

Reply:
Mostly quick prototyping of new implementations.
Moreover, in this case I did not want to depend on the segment implementation, that's why I made a way to create a new segment from another segment of any kind.

Reply:
Ok, I'll move everything back to the interface then. Just let me override frequently used properties in `Segment2DImpl` so as to avoid re-computation.

Reply:
How did you measure that the recomputation effort is greater than the cost in memory? And which is the impact on an average simulation?

Reply:
I didn't 😅 just assumed that the recomputation effort is significant as those properties are used massively, and that memory costs are always bearable. I made a few tests in the past hour: no measurable difference in recomputing those informations. I'm moving everything back to interface.

Reply:
Yes of course there's no difference, there are two compilers between your code and the performance, and most of the stuff is plain ALU :laughing: 

General approach: do the cleanest thing, measure performance, then act on what's holding them back.
By the way memory is most definitely not always bearable, not even within the scope of Alchemist, and actually is currently used in a quite suboptimal way in the simulator. However, the impact of this implementation would be negligible in any case.

Reply:
Thanks for the tips 😋